src/snowflake/snowpark/functions.py

```diff
@@ -8700,3 +8711,135 @@ def make_interval(
+        >>> df = session.create_dataframe([[10, "a"]], schema=["col1", "col2"])
+        >>> new_df = map(df, lambda row: row[0] * row[0], output_types=[IntegerType()])
+        >>> new_df.show()
+        ---------
+        |"C_1"  |
+        ---------
+        |100    |
+        ---------
+        <BLANKLINE>
+
+    Example 2::
+
+        >>> new_df = map(df, lambda row: (row[1], row[0] * 3), output_types=[StringType(), IntegerType()])
+        >>> new_df.show()
+        -----------------
+        |"C_1"  |"C_2"  |
+        -----------------
+        |a      |30     |
+        -----------------
+        <BLANKLINE>
+
+    Example 3::
+
+        >>> new_df = map(df, lambda row: (row[1], row[0] * 3), output_types=[StringType(), IntegerType()], output_column_names=['col1', 'col2'])
+        >>> new_df.show()
+        -------------------
+        |"COL1"  |"COL2"  |
+        -------------------
+        |a       |30      |
+        -------------------
+        <BLANKLINE>
+
+    Example 4::
+        >>> df = session.create_dataframe([10, 20, 30])
+        >>> new_df = map(df, lambda x: x * x, output_types=[IntegerType()], wrap_row=False)
+        >>> new_df.sort(new_df[0]).show()
+        ---------
+        |"C_1"  |
+        ---------
+        |100    |
+        |400    |
+        |900    |
+        ---------
+        <BLANKLINE>
+
+    Note:
+        The result of the `func` function must be either a scalar value or a tuple containing the same number of elements
+        as specified in the `output_types` argument.
+    """
+    if "packages" not in kwargs:
+        kwargs["packages"] = ["snowflake-snowpark-python"]
+    if len(output_types) == 0:
+        raise ValueError("output_types cannot be empty.")
+    input_types = [field.datatype for field in dataframe.schema.fields]
+    num_fields = len(dataframe.schema.fields)
+    output_schema = [
+        StructField(f"c_{i + 1}", output_types[i]) for i in range(len(output_types))
+    ]
+    output_cols = [
+        col(f"${i + 1}") for i in range(num_fields, num_fields + len(output_types))
+    ]
+    input_row_obj = snowflake.snowpark.Row(*dataframe.columns) if wrap_row else None
+
+    def wrap_result(result):
+        if isinstance(result, list) or isinstance(result, snowflake.snowpark.Row):
+            return tuple(result)
+        elif isinstance(result, tuple):
+            return result
+        else:
+            return (result,)
+
+    class _MapFunc:
+        def process(self, *argv):
+            if wrap_row:
+                argv = [input_row_obj(*argv)]
+            result = func(*argv)
+            yield wrap_result(result)
+
+    map_udtf = dataframe._session.udtf.register(
+        _MapFunc,
+        output_schema=StructType(output_schema),
+        input_types=input_types,
+        is_permanent=False,
+        **kwargs,
+    )
+
+    if not output_column_names:
+        output_column_names = [field.name for field in output_schema]
+
+    _output_cols = [
```

Comment:
If we implement output_column_names as done above, we can just do `dataframe.join_table_function(map_udtf(*dataframe.columns)).select(output_column_names)` in the final return

Reply:
@sfc-gh-aalam  the problem with selecting the columns by name (via the `output_column_names`created in the previous comment) is that we can get an ambiguos column reference. The reason for this is that the `join_table_function` includes the original columns of the DataFrame. The following scenarios illustrate this problem:

1. When chaining several `map` calls (because of the column names we generate):

```python
another_df = session.create_dataframe([('a',10), ('b', 20), ('d', 30)], 
                                      schema=StructType([StructField('name', StringType()),
                                                         StructField('numval', IntegerType())]))

r1 = map(another_df, lambda row: (row[0] + "_", row[1] * row[1]), [StringType(), IntegerType()])
r2 = map(r1, lambda row: row[0], [StringType()])
r2.show()
```

2. When processing a DataFrame that already has  `C1,C2,...` columns (or `C_1, C_2,...` before this change):

```python
another_df = session.create_dataframe([('a',10), ('b', 20), ('d', 30)], 
                                      schema=StructType([StructField('C1', StringType()),
                                                         StructField('C2', IntegerType())]))

r1 = map(another_df, lambda row: (row[0] + "_", row[1] * row[1]), [StringType(), IntegerType()])
r1.show()
```

In both these scenarios we get the following error if we apply this change:

```
SnowparkSQLException: (1304): 01b773a7-0a0c-34b2-0000-1c550d43832a: 002028 (42601): SQL compilation error:
ambiguous column name 'C1'
```

Reply:
make sense. We should  move this part in `_output_cols` initialization too.